### PR TITLE
feat: Tray icon voice detection

### LINKED ISSUE #294: [Feature Request] Restore Discord "Audio Detected" tray icon?
I like Discords tray icon when you're in a vc, I usually use that to tell if my voice is being picked up so I can have things open over discord, Vesktop seems to disable/replace that, and so far that's the only standard discord feature I miss (if it does exist I just cant find it). There was another issue thread where someone asked for custom tray icons and it was just closed without mention of that feature being added, so if it was added I can't find that either. 
Would be nice as a settings toggle or a plugin, I would even accept having to upload my own icons for the feature to work.

### LINKED ISSUE #667: User customizable assets
### Content

This is an issue which is meant to unify #571, #352 and #355, #439, #517, #576, #261 as well as #298, #424, #384, #294.
https://discord.com/channels/1015060230222131221/1216096162008924291/1252079440075161715

Design goals:
1. Give the user full control over what icons Vesktop displays and/or uses. 
      - excludes; in window assets, controllable via css and the app icon(read #505, DEs provide methods of achieving this). 
2. Store the user provided assets in the Vesktop cache dir. (all the prs that partially implement things do not do this.)
3. Support Linux, Windows and MacOS unless stated otherwise

Scope:
1. Startup splash customization
2. Tray icon customization
 
Features:
1. [x] Startup Splash: 
(mostly implemented in #355)
     - [x] Support static and animated image formats
     - [ ] Multi selection for randomized startup animation
     - [x] Disable the startup splash window in its entirety #384 (Done in #820)
     - [x] Adhere to design goal `#2`

2. [x] Tray icon:
Implemented in #517 
      - [x] Change static icon
           1. [x] User can provide a static icon
           2. [x] Support unread badges:
                - Bake the unread indicator into the tray icon #261. (Colour is user pick-able or follows system  accent)
     - [ ] Voice chat status
          1. [ ] Default audio, mute and deafened indicator follow system accent colour unless specified otherwise by a provided option 
          2. [ ] Default light and dark mode variants of the above which automatically switch using [this](https://www.electronjs.org/docs/latest/api/native-theme#properties) unless overridden by the user
          3. [ ] Allow the user to apply custom icons for any of the above 4 states (Audio detected, no audio, muted, deafend)

<details>
     <summary>FYI</summary>
     System accent on linux is a standard portal. Read: flatpak/xdg-desktop-portal/pull/815. Try in your terminal:
     
    dbus-send --session --print-reply --dest=org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop org.freedesktop.portal.Settings.Read string:'org.freedesktop.appearance' string:'accent-color'

</details>

cc: @Vendicated @PolisanTheEasyNick @ading2210 @alchzh @MrGarlic1

### PR DESCRIPTION
**Pull request by PolisanTheEasyNick**
epic pr🚀 cooking explodin
Resolves #294, partially resolves #667
- [x] Show icons depending on current voice state

Any ideas about improving PR structure are welcome ^^

---

### Voice Detection Icons
image image image image

---

### Notification Badge on top of Tray Icon
image

Special thanks to @MrGarlic1 and @Covkie 

Also, on my repo, there is branch `tray-icon-old` where all changes with tray icons customizations, colors, and so on

### CODE REVIEW

**Comment by PolisanTheEasyNick** [comment]
Thank you for feedback! I'm not using any black colors, I believe that it was just small border width. Also, I increased whole icon size. Can you please pull and retest?
What state is on your last screenshot? If it is muted or deafened that means that icons too small, and we need to redraw them completely, because I just took discord's original icons and changed colors to Vesktop ones. They are 24x24 pixels, so maybe they have not resized to proper size in Gnome system.

**Comment by EtereosDawn** [comment]
> Thank you for feedback! I'm not using any black colors, I believe that it was just small border width. Also, I increased whole icon size. Can you please pull and retest? What state is on your last screenshot? If it is muted or deafened that means that icons too small, and we need to redraw them completely, because I just took discord's original icons and changed colors to Vesktop ones. They are 24x24 pixels, so maybe they have not resized to proper size in Gnome system.

I can confirm that it looks much better! Well, maybe the edges of the circle could be a little smoother but I believe that the pixels are due to my monitor. That's great :) thanks for accepting the feedbeck.

image

File: src/renderer/patches/tray.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    FluxDispatcher.subscribe("RTC_CONNECTION_STATE", params => {
+        if (params.state === "RTC_CONNECTED") {
+            if (deafActions.isSelfDeaf()) {
+                VesktopNative.app.setTrayIcon("deafened");
+            } else if (muteActions.isSelfMute()) {
+                VesktopNative.app.setTrayIcon("muted");
+            } else {
+                VesktopNative.app.setTrayIcon("idle");
+            }
```
Comment (Vendicated) [inline src/renderer/patches/tray.ts:58]: this if, else if, else block is duplicated 4 times. it should be extracted into a function

File: src/main/mainWindow.ts [hunk 2]
```diff
@@ -478,3 +478,23 @@ export async function createWindows() {
 
     initArRPC();
 }
+
+export async function setTrayIcon(iconName) {
```
Comment (Vendicated) [inline src/main/mainWindow.ts:482]:
```suggestion
export async function setTrayIcon(iconName: string) {
```

File: src/main/mainWindow.ts [hunk 3]
```diff
@@ -478,3 +478,23 @@ export async function createWindows() {
...
+    if (!tray) return;
+    switch (iconName) {
+        case "speaking":
+            tray.setImage(SPEAKING_ICON_PATH);
```
Comment (Vendicated) [inline src/main/mainWindow.ts:486]: whats the point of making all these constants? why not just
```js
const Icons = new Set(["speaking", "muted", "deafened", "idle", "main"]);
// ...

if (!Icons.has(iconName)) {
	console.warn("setTrayIcon: Invalid icon name", iconName);
	iconName = "main";
}
tray.setImage(join(ICON_DIR, iconName + ".png"));
```

**Comment by PolisanTheEasyNick** [comment]
Added ability to change colors. For now, I want to find a way how to disable icon color chooser in settings when tray icon is disabled instead of hiding the whole setting and (maybe) add ability to propose to user his system accent color.

File: src/renderer/components/settings/TrayColorPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { SettingsComponent } from "./Settings";
+
+const ColorPicker = findByCodeLazy(".Messages.USER_SETTINGS_PROFILE_COLOR_SELECT_COLOR", ".BACKGROUND_PRIMARY)");
+
+const presets = [
+    "#3DB77F", // discord default ~
+    "#F6BFAC" // Vesktop inpired
+];
+
+export const trayIconPicker: SettingsComponent = ({ settings }) => {
```
Comment (Vendicated) [inline src/renderer/components/settings/TrayColorPicker.tsx:23]: react components should be PascalCase

File: src/renderer/components/settings/TrayColorPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    if (!settings.tray) return null; // how to disable instead of hiding?
+    return (
+        <div className="tray-settings">
+            <div className="tray-container">
+                <div className="tray-settings-labels">
```
Comment (Vendicated) [inline src/renderer/components/settings/TrayColorPicker.tsx:28]: we prefix our classes to prevent potential conflicts

vencord usually uses `.vc-*` whereas vesktop usually uses `.vcd-`

File: src/renderer/components/settings/TrayColorPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+const ColorPicker = findByCodeLazy(".Messages.USER_SETTINGS_PROFILE_COLOR_SELECT_COLOR", ".BACKGROUND_PRIMARY)");
+
+const presets = [
+    "#3DB77F", // discord default ~
+    "#F6BFAC" // Vesktop inpired
+];
+
+export const trayIconPicker: SettingsComponent = ({ settings }) => {
+    if (!settings.tray) return null; // how to disable instead of hiding?
```
Comment (Vendicated) [inline src/renderer/components/settings/TrayColorPicker.tsx:24]: is this actually necessary to change? returning null is the correct way in react to render nothing

Comment (PolisanTheEasyNick) [reply src/renderer/components/settings/TrayColorPicker.tsx:24]: But in settings we also have "Minimize to tray" which disables when Tray is off and I wanted to achieve similar behavior for this too. Because one element just *grays* as inactive and another (mine) disappearing at all

**Comment by PolisanTheEasyNick** [comment]
> displays the speaking/audio detected icon despite no user-speaking occurring.

Fixed in latest commit)
Interesting thing is that Discord calls SPEAKING event even when sound outcoming from video stream, so I've added additional checks for event call context.

**Comment by D3SOX** [comment]
Great work! Can you add an option to disable the badge or have it only for pings? I find it very annoying because I pretty much always have some unread messages

File: src/main/tray.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    return Settings.store[settingKey];
+};
+
+export async function setTrayIcon(iconName: string) {
+    if (!tray || tray.isDestroyed()) return;
+    const Icons = new Set(["speaking", "muted", "deafened", "idle", "icon"]);
+    if (!Icons.has(iconName)) return;
+
+    // if need to set main icon then check whether there is need of notif badge
+    if (iconName === "icon" && lastBadgeCount > 0) {
```
[the remark this answers is not in the document]
Comment (D3SOX) [reply src/main/tray.ts]: I'm not sure if I understand the new behavior correctly. After updating, it now shows the dot for non-mentions again.
Am I supposed to disable “Enable Unread Message Badge” when I want it to only for mentions?
image

I think this could be made clearer

**Comment by alchzh** [comment]
> Might it be best to fold #261 into this PR?

I just found out about this PR, looks great though I haven't tested it yet. This only shows a notification dot not a number, right? if there's interest in getting numbers in the tray icons like I have it in that PR i'm happy to work on top of your PR @PolisanTheEasyNick (it's an important feature for me)

**Comment by PolisanTheEasyNick** [comment]
> > Might it be best to fold #261 into this PR?
> 
> I just found out about this PR, looks great though I haven't tested it yet. This only shows a notification dot not a number, right? if there's interest in getting numbers in the tray icons like I have it in that PR i'm happy to work on top of your PR @PolisanTheEasyNick (it's an important feature for me)

I can add that easily (even already did once at [c725352](https://github.com/Vencord/Vesktop/pull/517/commits/c725352bef0dd85f110956593e8a5a79a540d900) and removed at [8bbe16b](https://github.com/Vencord/Vesktop/pull/517/commits/8bbe16b775a6b5b9b2ac6fd8ec168a6cf69fc8e5)), but problem was that numbers on badge in tray icon is very small and not readable, so there was a decision in ds chat to show only notification dot.
Also, this PR is already quite massive and badge logic on Linux is quite overwhelming, so probably the best approach here will be to rework the whole notification badge settings and with this add support of enabling/disabling badge numbers in tray icon :)

**Comment by zoobporsor** [comment]
Where is the option to change the taskbar and tray icon? I cant find it

**Comment by PolisanTheEasyNick** [comment]
> Where is the option to change the taskbar and tray icon? I cant find it

Vesktop Settings -> Tray  
image
If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch

**Comment by prochy-exe** [comment]
Hello, i tried building your branch for vencord however i get and error endlessly and the icon doesnt change what so ever.
Vesktop_2024-08-27_14-06-14

Apart from that this is also being spammed in the console. That's all the useful info i could find.
region_2024-08-27_14-07-37

I am running this on arch, running hyprland.

**Comment by PolisanTheEasyNick** [comment]
> I tried running this on windows and it works just fine, the only problem is that the icon is not customizable 

Try retoggle "Enable tray icon setting"   

> The only problem is that it looks a bit strange? I'm not sure if this is visible on the screenshot

It is strange indeed. Looks like some problem with icon sizes on Windows :c

**Comment by zoobporsor** [comment]
> > Where is the option to change the taskbar and tray icon? I cant find it
> 
> Vesktop Settings -> Tray image If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch

Ok but this doesnt exist in the main .app image? I need to download some special pre-release build?  Its not out yet officially?

**Comment by PolisanTheEasyNick** [comment]
> Ok but this doesnt exist in the main .app image? 

No, it will be at .app when this PR would be merged and new Vesktop release released
> I need to download some special pre-release build? Its not out yet officially?

You can clone `tray-icon` branch from my fork and start Vesktop from console

**Comment by prochy-exe** [comment]
> > > Where is the option to change the taskbar and tray icon? I cant find it
> > 
> > 
> > Vesktop Settings -> Tray image If you still can't find try retoggle "Enable Tray Icon" option. If still can't find then make sure you're using `tray-icon` branch
> 
> Ok but this doesnt exist in the main .app image? I need to download some special pre-release build? Its not out yet officially?

i made this repo so if you trust me enough you can get it from here, if you are not on mac that is https://github.com/prochy-exe/custom-vesktop-builds

**Comment by BreadYogurt** [comment]
Using SVG files doesn't seem to scale properly for me, here's the results from trying to use the discord tray icons included in the Papirus icon set.
Screenshot_20240903_093546

**Comment by PolisanTheEasyNick** [comment]
Can you please send .svg file you're trying to set? I'll try to test and fix ASAP

**Comment by BreadYogurt** [comment]
Here it is.
discord-tray

**Comment by BreadYogurt** [comment]
Yep, that fixed it, and more importantly, nothing exploded!

File: src/renderer/components/settings/TraySettings.tsx [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+import { Margins, Modals, ModalSize, openModal } from "@vencord/types/utils";
+import { findByCodeLazy, findByPropsLazy } from "@vencord/types/webpack";
+import { Button, Forms, Select, Switch, Toasts } from "@vencord/types/webpack/common";
+import { setCurrentTrayIcon } from "renderer/patches/tray";
+import { useSettings } from "renderer/settings";
+
+import { SettingsComponent } from "./Settings";
+
+const ColorPicker = findByCodeLazy(".Messages.USER_SETTINGS_PROFILE_COLOR_SELECT_COLOR", ".BACKGROUND_PRIMARY)");
```
[the remark this answers is not in the document]
Comment (PolisanTheEasyNick) [reply src/renderer/components/settings/TraySettings.tsx]: Fixed in 4bd812a, thanks @Covkie!

**Comment by mrbigmouth502** [comment]
The lack of a system tray notification dot in Vencord/Vesktop is the #1 thing keeping me from switching to it from the stock Discord client. It's an extremely important feature for my workflow. 

Hopefully the conflicts preventing it from being merged can be sorted out. 🙂

**Comment by Thespikedballofdoom** [comment]
> The lack of a system tray notification dot in Vencord/Vesktop is the number one thing keeping me from switching to it from the stock Discord client. It's an extremely important feature for my workflow.
> 
> Hopefully the conflicts preventing it from being merged can be sorted out. 🙂

Same, really, really pining for this, it's the only thing I look for every update.

**Comment by Vendicated** [comment]
Sorry for the long wait

I really love the addition of the tray icon modifiers, thank you!

Some feedback:
1. Please resolve conflicts
2. You seem to have added a native dbus library. Do we really need this? If yes, make sure that it is correctly bundled into the binary when packaging, not just when running dynamically from source. Also make sure this won't cause issues when running on non-linux, and make sure arm64 is correctly supported
3. I'm not really a fan of the customisations being baked into this PR. I do plan to have one central UI for managing various images but that should be part of a separate change, but the way you implemented it here doesn't really match what I had envisioned. Please remove the image customisation logic & ui. That will drastially reduce the complexity of this change

In other words, do keep all the tray modifiers like showing voice status, notification dot etc, but remove the settings ui

**Comment by kerriganx** [comment]
It is possible to add ability to set custom icon for unread messages instead of red dot? Thanks

**Comment by PolisanTheEasyNick** [comment]
Thank you for response!

> You seem to have added a native dbus library. Do we really need this? 

I'm using it for getting accent color from Linux system. The library was added because it may be used for some other features, like #686.  
Personally, I think that feature "automatically pick accent color and propose it to user" is not very critical, because user can just pick color by himself, so I will remove library and accent color feature from this PR scope.

> but remove the settings ui

Should I remain color picker settings?

**Comment by KarlOfDuty** [comment]
> I do plan to have one central UI for managing various images but that should be part of a separate change

Is this planned soon? If its like a year away couldn't this gui be allowed to exist until then? It would be a shame to have these features ready to go but then get removed until really far into the future.

**Comment by fragtion** [comment]
> 3. I'm not really a fan of the customisations being baked into this PR. I do plan to have one central UI for managing various images but that should be part of a separate change, but the way you implemented it here doesn't really match what I had envisioned. Please remove the image customisation logic & ui. That will drastially reduce the complexity of this change
> 
> In other words, do keep all the tray modifiers like showing voice status, notification dot etc, but remove the settings ui

How should the icons now be modified now that the UI is removed? The default accent color is not striking/noticeable enough for me and I would like to be able to change it like we could earlier in this PR, but there seems to be no obvious way of doing that anymore...? Is this just a temporary regress?

**Comment by mrbigmouth502** [comment]
> This branch is out-of-date with the base branch
> 
> Changes can be cleanly merged.

Hmm...

**Comment by Vendicated** [comment]
I just added tray customisation via #1179 so if you rebase on that, you can add customisation back!

Getting the system accent colour will be made possible via #1180 so then we can just use the user's system accent colour for the badge modifier colour

**Comment by mrbigmouth502** [comment]
I just wanna say thanks to whoever added tray notification badges recently. It's allowed me to finally switch over from the stock Discord client. 🙂

**Comment by PolisanTheEasyNick** [comment]
Hi @Vendicated,

I have updated branch to support newest master.
Currently this PR provides only support for voice tray icons.
Regarding changing badge color(s) according to system's accent color: it is possible, but wil greately overwhelm this PR, because images pre-processing should be done according to system accent color.
I think, that in scope of this PR it is okay to implement only tray icon voice detection, and, if needed - changing colors based on system accent in another PR.
But, I also think that it is not worth to do because of support of setting custom images for icons.
Waiting for a review and will be happy to fix issues.
TIA!
